### ISSUE TITLE
Question: Using distributed instances, targeting a single user for emit from outside flask

### ISSUE THREAD

**Issue by jdu** [miguelgrinberg/Flask-SocketIO#301]
I'm currently using flask-socketio in a fairly sizeable app for the humanitarian sector and I am having a slight problem finding the best way to handle a fairly specific problem. 

I'm running multiple instances of the socket application using haproxy to stick sessions for a given socket as well as the multiple instances are tied together with rabbitmq. 

The problem I have is that we have some disconnected services as well as background worker processes which I'd like to allow emitting to a specific user or group of users from said process. Some of these processes aren't in Python. These other services won't have access to the request.sid of a specific user only their I'd from our system. 

Right now what I have is on connect in the socket application, I grab the sid and our own internal session storage and register the user into a dict where the keys are their actual user id and the values are their sid. On disconnect I remove them. 

What I want to know is what would be the recommended approach for allowing a completely disconnected process in potentially a different language and even potentially separate server (with access to rabbitmq) which only has knowledge of the users internal application Id, the ability to push a user targeted message through to flask-socketio to emit to the user on the appropriate running instance of the socket server?

I'm just interested in what your recommendations are or even potentially modification points in the package that would be the best place to customise flask-socketio to support this.

**Comment by miguelgrinberg** [comment]
This is the kind of problems I enjoy discussing. :)

I'm using the facilities in the [kombu](https://github.com/celery/kombu) package to post stuff to RabbitMQ. I really haven't looked in detail what kombu does when it writes to the queue, but if you need to post to the queue from other languages, all that needs to be done is to replicate the same thing kombu does. So the task here is to figure out exactly what operation(s) Kombu does to post to the RabbitMQ queue.

As far as the actual payload, you can look in file [pubsub_manager.py](https://github.com/miguelgrinberg/python-socketio/blob/master/socketio/pubsub_manager.py), specifically in the `emit()` method, for what you need to post. There are two formats for this payload, one is binary, based on the "pickle" package from the Python library, and the other is json. In my code I send payloads with the pickle format, but on the receiving side I accept both pickle and json. Clearly from a non-python process json is your only option.

If you look through the code you'll see there is the Kombu code and there is also code to talk directly to a Redis store, via the pubsub functions. For the Redis code, I have been able to post to the queue from a non-python process, just post the json payload to the pubsub channel and you're done.  So this should be totally doable for RabbitMQ as well.

If you end up implementing some sort of wrapper in another language that can emit via a Flask-SocketIO server, I would be super interested in hosting it in this repo for other users to have.

**Comment by jdu** [comment]
I'd not heard of Kombu before now. We're using Pika and various other methods for some of the other languages (C++, Go and Java). 

Looking at the Kombu Client Manager, I think I could get away with using a modified version. Essentially I could do is modify the KombuManager to override `connect` to retrieve data from the session when the user connects and store it in a dict in the class mapping user_id -> sid. 

Then on the other side of things, I need to set up my publishers to send a modified payload. What I'll probably do is write a thin wrapper around an AMQP lib for each language which provides `.emit()`, an `emit` method only, with an option to provide additional data. So for examples sake in Python, I would have a light-weight wrapper which allowed me to call `emit("event", data, properties=dict(user_id: 1))` at least with RabbitMQ (not sure about other messaging platforms) you can add custom BasicProperties to the message being pushed to the queue, for instance you can add content_type to explicitly state the message body format is JSON, XML, etc... I can also add in the target user_id or user_group as a property. 

```
properties = pika.BasicProperties(user_id='1',
                                          content_type='application/json',
                                          headers=message)

self._channel.basic_publish(self.EXCHANGE, self.ROUTING_KEY,
                                    json.dumps(message, ensure_ascii=False),
                                    properties)
```

On the other end, I can just modify the _listen method of the KombuManager to retrieve the user_id from the message properties and simply do ` if user_id in self.sid_uid_map.keys()` and if it comes back true create a message payload targeting the sid and yield it from KombuManager. If there's no match on user_id don't yield anything. Targeting a group id would be similar in that I map sids into a dict where the keys represent the group types, and in the connect override create an additional call to self.enter_room declaring a room for the group type. Then when yielding I yield to that groups room.

Putting the user_id, group_id, etc... in BasicProperties means I don't end up mucking about having to make a different message format and plucking/modifying it before yield, if there's no user_id, just yield the message (i.e. interpret it as a broadcast).

What do you think? Any visible choke points? it's a bit rough, but it seems like it could work alright. I can definitely share over any wrappers I make for the other languages as well as the base wrapper I set up for python to do this.

**Comment by miguelgrinberg** [comment]
> Essentially I could do is modify the KombuManager to override connect to retrieve data from the session when the user connects and store it in a dict in the class mapping user_id -> sid

You have to keep in mind that this is not a one to one mapping. A user could have multiple connections, so for a given user id you may have more than one `sid` value.

Is there really a benefit in moving the user_id to sid mapping to the manager class? You can do this easily in your application, when you listen to connect and disconnect events. You are probably doing that already, right?

> Then on the other side of things, I need to set up my publishers to send a modified payload.

To me, this does not sound like the best implementation. If I had to do this, I would keep the emit() function and the payloads the same. I think it could make sense for you to write a new client manager instance that uses Pika instead of Kombu, as that will allow you to have control over the way the payload is posted to the queue. But I would not include user ids in the payload, to me it seems cleaner to keep the mapping of user ids to sids in the application space. Any process that wants to post to the queue can do so by obtaining the sid of the desired user from your session storage. Your wrapper functions for each language could do that, actually, so at the high level application layers you are using user ids. Then this solution is portable, if one day you want to switch to Redis for example, it's zero effort, just switch to my Redis manager class and you're done.

Also consider that having the user id to sid map in a Python memory structure does not scale. The day you need to run two server processes that will break.

**Comment by jdu** [comment]
True, I completely forgot that a user could have multiple connections open at any given time.

And I actually misconstrued the usage of client_manager. I think the naming was a bit misleading as I was thinking "manager of connected clients" (as well as it was a bit late in the evening and I was a bit over-tired hah). The managing of mappings from users to sid(s) can stay in the connect/disconnect events in the main application for sure.

> But I would not include user ids in the payload, to me it seems cleaner to keep the mapping of user ids to sids in the application space. Any process that wants to post to the queue can do so by obtaining the sid of the desired user from your session storage. Your wrapper functions for each language could do that, actually, so at the high level application layers you are using user ids. Then this solution is portable, if one day you want to switch to Redis for example, it's zero effort, just switch to my Redis manager class and you're done.

This seems actually less portable to me, because you are now persisting state from the socket application into the larger application space. In my case as well, we're using AES 256 encrypted cookie storage for sessions at the moment (the decision to do this was mainly to expedite development and it will be switching to db-backed session storage later on) which means I don't have an easily accessible session storage to persist the sids for retrieval elsewhere. 

I had a think about this over the weekend (got your response Sat but was busy all weekend.) and I think the wrappers would definitely need to be suitably generalized to not include parameters different from the API defined for emit in the python-socketio implementation. However, in my case, where I don't want sids polluting the application space and because flask-socketio in my case acts as an interface to core logic and decoupled daemons/services, and needs to swappable out for different types of interfaces (REST/HTTP, UDP, TCP, etc...) I would need to have some type of authority that is responsible (only in my case) for maintaining a list of connected user ids against their sid(s) that is specific to the flask-socketio context. Right now I don't want to implement another backend component to support this, however, others might, so to support that I think the best way to achieve it would be to allow the wrapper class in each specific language to be sub-classable and the emit method overrideable to be customised for a specific use-case, the default method signature being the same as the python-socketio emit signature.

I have completely disconnected daemon workers who have restricted access to specific tables. For instance, one has access to a table which contains a user_id column, it's never exposed to any type of HTTP request and can't access the cookie for the user. So it would need to be able to send an emit based on purely the user_id and I would rather it stay that way so that the interface is purely responsible for managing its domain. 

The other thing I could do (and I think I will do for the interim), is to pass user_id as the room to the emit -> room parameter in my wrapper class, keep my users -> sids mapping in the flask-socketio apps like I said for now as a dict (user_id, [...sids...]) and then with my ClientManager class, I pass in that dict as part of __init__ and then in the `_listen` method I simply modify the message swapping out the user_id for the appropriate sids before I yield the message. 

That way I'm creating a generalised wrapper in the other languages still that can work for other peoples use cases while allowing sub-classing in order to customize it for their specific needs, and the ability to sub-class the client_manager allows them to customize it on the flask-socketio side for their specific needs. It also means that further down the road, once I have persistence-backed sessions in place all I have to modify is the flask-socketio interface side storage of the sids to update the persisted session data and the client-manager retrieval of that and nothing else. Even the message queue can stay the same as your KombuManager and the implementation is fairly simple to set up in an example application. This means the WebSocket interface (i.e flask-socketio) is concerned only with what it needs to be concerned with and the application logic doesn't need to know anything about the interface that it's interacting with (i.e. the sids).

When I do have persistance backed session storage, it will have to be on something that supports high-speed writes/reads in order to support the number of clients connecting/disconnecting and scale appropriately for larger loads, i'll have to have a think about that. Storing it in memory in a dict() isn't the most performant way to achieve it but for now it will have to do until I have persisted sessions.
